File: 리트코드/0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {

        List<List<Integer>> result = new ArrayList<>();
        dfs(result, nums, new ArrayDeque<>(), 0);
        
        return result;
    }

    private void dfs(List<List<Integer>> result,int[] nums, Deque<Integer> path, int index) {
        result.add(path.stream().collect(Collectors.toList()));

        for (int i = index; i < nums.length; i++) {
            path.add(nums[i]);
            dfs(result, nums, path, i + 1);
            path.removeLast();
        }
    }
}

File: 리트코드/0078-subsets/0078-subsets.kt
class Solution {
    fun subsets(nums: IntArray): List<List<Int>> {
        val result: MutableList<List<Int>> = mutableListOf()
        fun dfs(index:Int, path:ArrayDeque<Int>) {
            result.add(path.stream().collect(Collectors.toList()))

            for (i in index until nums.size) {
                path.add(nums[i])
                dfs(i + 1, path)
                path.removeLast()
            }
        }

        dfs(0, ArrayDeque())
        return result
    }
}